docs/reference/api-concepts.md

```diff
@@ -0,0 +1,232 @@
+
+* Cluster-scoped resources: 
+  * `GET /apis/GROUP/VERSION/RESOURCETYPE` - return the collection of resources of the resource type
+  * `GET /apis/GROUP/VERSION/RESOURCETYPE/NAME` - return the resource with NAME under the resource type
+* Namespace-scoped resources: 
+  * `GET /apis/GROUP/VERSION/RESOURCETYPE` - return the collection of all instances of the resource type across all namespaces
+  * `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE` - return collection of all instances of the resource type in NAMESPACE
+  * `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME` - return the instance of the resource type with NAME in NAMESPACE
+
+Since a namespace is a cluster-scoped resource type, you can retrieve the list of all namespaces with `GET /api/v1/namespaces` and details about a particular namespace with `GET /api/v1/namespaces/NAME`.
+
+Almost all object resource types support the standard HTTP verbs - GET, POST, PUT, PATCH, and DELETE. Kubernetes uses the term **list** to describe returning a collection of resources to distinguish from retrieving a single resource which is usually called a **get**. 
+
+Some resource types will have one or more sub-resources, represented as sub paths below the resource:
+
+* Cluster-scoped subresource: `GET /apis/GROUP/VERSION/RESOURCETYPE/NAME/SUBRESOURCE`
+* Namespace-scoped subresource: `GET /apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME/SUBRESOURCE`
+
+The verbs supported for each subresource will differ depending on the object - see the API documentation more information. It is not possible to access sub-resources across multiple resources - generally a new virtual resource type would be used if that becomes necessary.
+
+
+## Efficient detection of changes
+
+To enable clients to build a model of the current state of a cluster, all Kubernetes object resource types are required to support consistent lists and an incremental change notification feed called a **watch**.  Every Kubernetes object has a `resourceVersion` field representing the version of that resource as stored in the underlying database. When retrieving a collection of resources (either namespace or cluster scoped), the response from the server will contain a `resourceVersion` value that can be used to initiate a watch against the server. The server will return all changes (creates, deletes, and updates) that occur after the supplied `resourceVersion`. This allows a client to fetch the current state and then watch for changes without missing any updates. If the client watch is disconnected they can restart a new watch from the last returned `resourceVersion`, or perform a new collection request and begin again.
```

Comment:
> restart a new watch from the last returned `resourceVersion`

Ooh, is this official? :scroll: :crown:
We've bumped into how to restart watch from same point in ruby kubeclient (https://github.com/abonas/kubeclient/pull/275) and I see python client too (https://github.com/kubernetes-incubator/client-python/issues/124)...

So far all I read said `resourceVersion` is opaque and it wasn't clear they come from same timeline for different objects and their collection...
Unlike List that returns both collection resourceVersion and individual objects' versions, during watch you only see updates to individual objects' versions.

I see experimentally that the collection's resourceVersion as returned by List is *de-facto* same for all collections, increments on any change to object in *any* collection, and equals resourceVersion of that last changed object, so this would work.

So, can clients assume collection resourceVersion >= max(obj.resourceVersion for obj in collection)?
Can clients assume watching collection from max(obj.resourceVersion seen during previous watch) will yield same point watch stopped?  (as long as that history is not lost...)

More things that would be great if docs explicitly allowed or disallowed:
- Can clients take a collection version (from List) and use it on a single resource watch?
- Can clients take version from one collection and use it on another collection?
- Can clients do any kind of "greater than" comparisons between versions?
  Or only know that last string seen, over single watch, is semantically latest?

Reply:
> Ooh, is this official?

Yes.

> We've bumped into how to restart watch from same point

You can always attempt to restart from the last resourceVersion you received. If the api server no longer has enough history to let you start from that point, it will return a 410 error and you'll have to relist to get a new fresh resourceVersion

> So far all I read said resourceVersion is opaque and it wasn't clear they come from same timeline for different objects and their collection...

That is correct. You should not make any assumptions about how two resourceVersions relate to each other.

> I see experimentally that the collection's resourceVersion as returned by List is de-facto same for all collections, increments on any change to object in any collection, and equals resourceVersion of that last changed object, so this would work.

Things you can do:
* Remember the last resourceVersion you received for a resource, and use it when re-establishing a watch for that resource
* Compare the resourceVersion for an object retrieved at two different times to see if it is identical

Things you cannot do:
* Assume the resourceVersion is numeric
* Assume you can derive meaning from two different resourceVersions (e.g. "bigger number came later")
* Compare resourceVersions between two resources, or between two resource types

> So, can clients assume collection resourceVersion >= max(obj.resourceVersion for obj in collection)?

No

> Can clients assume watching collection from max(obj.resourceVersion seen during previous watch) will yield same point watch stopped? (as long as that history is not lost...)

You should not take the max... that involves interpreting the resourceVersion. You should treat it as opaque and remember the most recent one received

Reply:
@smarterclayton maybe you can incorporate some of the below, or do we have a canonical place for this to be documented to link to?

@cben, answers below.

> So, can clients assume collection resourceVersion >= max(obj.resourceVersion for obj in collection)?

Per collection, this happens to be true today. However the *only* operation on resourceVersions that we're promising will work in the indefinite future is that of equality (==). One can imagine data storage techniques where resourceVersions aren't linear or aren't numeric. Please don't bake assumptions about resourceVersions into clients, that's not future proof.

> Can clients assume watching collection from max(obj.resourceVersion seen during previous watch) will yield same point watch stopped? (as long as that history is not lost...)

Do not use max; use last seen. Then, yes, that is the way it is intended to function.

> Can clients take a collection version (from List) and use it on a single resource watch?

Allowed.

> Can clients take version from one collection and use it on another collection?

NO. Empirically it works... until it doesn't. The cluster administrator can choose to move resource types between different backends, which would cause this to stop being true. The default setup for Events already does this.

> Can clients do any kind of "greater than" comparisons between versions?

NO.

> Or only know that last string seen, over single watch, is semantically latest?

Probably not safe depending on what you want to do, in that there could be things you haven't yet seen over the watch.

Reply:
Thanks! <3

> Things you can do:
> - Remember the last resourceVersion you received for a resource, and use it when re-establishing a watch for that resource

Right, for watch on single resource this was clear.

Just to emphasize, I'm asking about watch of a *collection*, and restarting it from the last version of an *individual resource* within that collection — which is all you get as you watch.

Reply:
@lavalamp hello, a quick question: so today, we can say for EntityA, if it has bigger resourceVersion, it is newer (due to etcd's modifiedIndex).

So for the future, will there always be some way to say which version of the EntityA is newer?Because we should have such attribute. I would guess that many applications have this need.

Reply:
> today, we can say for EntityA, if it has bigger resourceVersion, it is newer (due to etcd's modifiedIndex).

that is not part of the API guarantee

> will there always be some way to say which version of the EntityA is newer?

there is creationTimestamp on the object, but other than that, no.

Reply:
> if we combine the data from API and watches, the data from watches can temporarily give us old data, that we want to throw away, but we can't because we don't know what is newer.

combining data from two streams that differ in time seems problematic. Typically, controllers are driven by watch alone, and if updates to the API hit a conflict error, that means the resource was updated in the meantime, and the controller can simply wait for the next event to arrive via the watch stream

Reply:
> we want to process the data in parallel, without something comparable, we are forced to do everything in 1 process

the workqueue used by most of the kubernetes controllers allows parallel processing of watch events... seeing how that is structured might be helpful (https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go#L131-L133)

Reply:
@liggitt Thank you, I'll try to check it out, although my Go skills are not great :-)

In detail what we want to do:

We fetch and save k8s data for various purposes (reporting, chargeback, etc.) into our Postgre SQL DB.

Given we have e.g. 100k pods and the app was down for a while (therefore the watches history is missing), we'd like to fetch it and start watches at the same time. So we get fresh changes immediately, while we are getting the full inventory on the background (saving in all in parallel into our DB)

Fetching 100k pods and associated objects and storing them takes quite some time, so if we need to do this sequentially, the user needs to wait to get fresh changes (in bigger envs it's like 0.5h or more of processing time). Also the sequential processing needs more orchestration and that makes it more complex.

There is no way of saving the data in parallel, if we can't compare the version of entities. So what would you advise here? I am not sure if kubernetes controllers are solving the same issue as described ^ ?

-----------------

So API and watches are taking the data from different sources? So if we would put e.g. the timestamp to k8s db for pods, the watches would not see it? Only the API query? Or what is the reason we can't have a comparable attribute?